homeassistant/components/azure_devops/__init__.py

```diff
@@ -0,0 +1,123 @@
+"""Support for Azure DevOps."""
+import logging
+from typing import Any, Dict
+
+from aioazuredevops.client import DevOpsClient
+import aiohttp
+
+from homeassistant.components.azure_devops.const import (
+    CONF_ORG,
+    CONF_PAT,
+    CONF_PROJECT,
+    DATA_AZURE_DEVOPS_CLIENT,
+    DATA_ORG,
+    DATA_PROJECT,
+    DOMAIN,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up the Azure DevOps components."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up Azure DevOps from a config entry."""
+    client = DevOpsClient()
+
+    try:
+        if entry.data[CONF_PAT] is not None:
+            await client.authorize(entry.data[CONF_PAT], entry.data[CONF_ORG])
+            if not client.authorized:
+                _LOGGER.warning(
+                    "Could not authorize with Azure DevOps. You may need to update your token."
+                )
+                raise ConfigEntryNotReady
```

Comment:
If this is an unrecoverable error, like failed authorization, we should not raise `ConfigEntryNotReady` as it will not improve the situation.

Can this be recovered without user interaction?

Normally we want to start a reauth config flow if a token needs to be updated and that requires user interaction.

Reply:
Well, this is more an 'I've deleted or regenerated my token in devops' situation. I'll look into triggering a reauth flow

Reply:
Here's an example:
https://github.com/home-assistant/core/blob/29ab1935cb78344511c2ddb1d9a8f413902c061c/homeassistant/components/samsungtv/media_player.py#L110-L111

https://github.com/home-assistant/core/blob/29ab1935cb78344511c2ddb1d9a8f413902c061c/homeassistant/components/samsungtv/config_flow.py#L163